File: src/backend/ai_chat_system.py
import requests
import json
from typing import List, Dict, Any
from config import config_manager
from search_api import SearchAPISystem

class AIChatSystem:
    """AI聊天系统核心类"""
    
    def __init__(self):
        """初始化AI聊天系统"""
        self.chat_history: List[Dict[str, str]] = []
        self.personality = config_manager.get_personality_config()
        # 限制聊天历史长度，避免API调用失败
        self.max_history_length = 20
        # 初始化搜索系统
        self.search_system = SearchAPISystem()
    
    def add_message(self, role: str, content: str) -> None:
        """添加消息到聊天历史"""
        self.chat_history.append({"role": role, "content": content})
        # 限制聊天历史长度
        if len(self.chat_history) > self.max_history_length:
            self.chat_history = self.chat_history[-self.max_history_length:]
    
    def clear_history(self) -> None:
        """清空聊天历史"""
        self.chat_history = []
    
    def get_chat_history(self) -> List[Dict[str, str]]:
        """获取聊天历史"""
        return self.chat_history
    
    def generate_response(self, user_input: str) -> str:
        """生成AI回复"""
        try:
            # 检查是否需要搜索
            search_needed = self._needs_search(user_input)
            search_results = ""
            
            if search_needed:
                # 使用Kimi执行搜索
                search_results = self.search_system.search(user_input)
                # 完全隐藏搜索结果，不打印任何信息给用户
                pass
            
            # 构建Deepseek的输入，包含用户问题和搜索结果
            if search_results and "搜索失败" not in search_results:
                deepseek_input = f"用户问题: {user_input}\n\n搜索结果:\n{search_results}\n\n请严格基于搜索结果回答用户的问题，不要添加任何搜索结果中没有的信息，保持自然友好的语气。如果搜索结果不完整或有问题，请如实告知用户。"
            else:
                deepseek_input = f"用户问题: {user_input}\n\n请如实告知用户你无法获取实时信息，因为你没有访问互联网的能力。建议用户使用其他方式查询相关信息，保持自然友好的语气。"
            
            # 清空聊天历史，避免重复
            self.chat_history = []
            self.add_message("user", deepseek_input)
            
            # 调用Deepseek API生成最终回复
            response = self._call_chat_api()
            
            return response
        except Exception as e:
            print(f"生成回复失败: {e}")
            return "抱歉，我暂时无法回答你的问题。"
    
    def _needs_search(self, user_input: str) -> bool:
        """判断是否需要搜索"""
        # 关键词列表，包含需要搜索的内容类型
        search_keywords = [
            "天气", "温度", "预报", "新闻", "资讯", "最新", "今天", "明天", "后天",
            "什么", "怎么", "如何", "为什么", "何时", "哪里", "谁", "哪个",
            "查询", "搜索", "查找", "了解", "知道", "想知道", "请问", "帮忙"
        ]
        
        # 检查用户输入是否包含搜索关键词
        user_input_lower = user_input.lower()
        for keyword in search_keywords:
            if keyword in user_input_lower:
                return True
        
        return False
    
    def _call_chat_api(self) -> str:
        """调用聊天API"""
        chat_config = config_manager.get_chat_api_config()
        api_key = chat_config.get("api_key")
        base_url = chat_config.get("base_url")
        model = chat_config.get("model", "gpt-3.5-turbo")
        temperature = chat_config.get("temperature", 0.7)
        max_tokens = chat_config.get("max_tokens", 1024)
        
        # 构建系统消息
        system_message = self._build_system_prompt()
        
        # 构建请求消息
        messages = [
            {"role": "system", "content": system_message}
        ] + self.chat_history
        
        # 根据不同的API提供商选择不同的调用方式
        if "openai" in base_url.lower() or not base_url:
            return self._call_openai_api(api_key, base_url, model, messages, temperature, max_tokens)
        elif "deepseek" in base_url.lower():
            return self._call_deepseek_api(api_key, base_url, model, messages, temperature, max_tokens)
        else:
            # 默认使用OpenAI兼容接口
            return self._call_openai_api(api_key, base_url, model, messages, temperature, max_tokens)
    
    def _build_system_prompt(self) -> str:
        """构建系统提示词"""
        return f"你是{self.personality.get('name')}，{self.personality.get('description')}。你的性格特点是：{self.personality.get('tone')}。初次见面时的问候语是：{self.personality.get('greeting')}。请保持这个设定与用户交流。"
    
    def _call_openai_api(self, api_key: str, base_url: str, model: str, messages: List[Dict[str, str]], 
                        temperature: float, max_tokens: int) -> str:
        """调用OpenAI API"""
        try:
            import openai
            
            # 配置OpenAI客户端
            openai.api_key = api_key
            if base_url:
                openai.api_base = base_url
            
            # 调用API
            response = openai.ChatCompletion.create(
                model=model,
                messages=messages,
                temperature=temperature,
                max_tokens=max_tokens
            )
            
            # 返回回复内容
            return response.choices[0].message.content
        except ImportError:
            print("OpenAI库未安装")
            return "抱歉，OpenAI库未安装，无法提供服务。"
        except Exception as e:
            print(f"OpenAI API调用失败: {e}")
            return "抱歉，我暂时无法回答你的问题。"
    
    def _call_deepseek_api(self, api_key: str, base_url: str, model: str, messages: List[Dict[str, str]], 
                          temperature: float, max_tokens: int) -> str:
        """调用DeepSeek API"""
        headers = {
            "Content-Type": "application/json",
            "Authorization": f"Bearer {api_key}"
        }
        
        payload = {
            "model": model,
            "messages": messages,
            "temperature": temperature,
            "max_tokens": max_tokens
        }
        
        try:
            # 发送请求
            response = requests.post(
                f"{base_url}/chat/completions",
                headers=headers,
                json=payload,
                timeout=30
            )
            
            # 解析响应
            response_data = response.json()
            if "choices" in response_data and response_data.get("choices"):
                return response_data["choices"][0]["message"]["content"]
            else:
                print(f"DeepSeek API响应错误: {response_data}")
                return "抱歉，我暂时无法回答你的问题。"
        except Exception as e:
            print(f"DeepSeek API调用失败: {e}")
            return "抱歉，我暂时无法回答你的问题。"

# 创建全局AI聊天系统实例
ai_chat_system = AIChatSystem()

File: backend/src/main/java/com/llm/vtb/ai/AiChatSystem.java
package com.llm.vtb.ai;

import com.llm.vtb.config.ConfigManager;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.util.HashMap;
import java.util.Map;
import com.fasterxml.jackson.databind.ObjectMapper;

@Component
public class AiChatSystem {

    @Autowired
    private ConfigManager configManager;

    private final ObjectMapper objectMapper = new ObjectMapper();

    public String generateResponse(String userInput) {
        try {
            ConfigManager.Config config = configManager.getConfig();
            ConfigManager.ChatApiConfig chatApiConfig = config.getChat_api();

            String apiKey = chatApiConfig.getApi_key();
            String baseUrl = chatApiConfig.getBase_url();
            String model = chatApiConfig.getModel();
            double temperature = chatApiConfig.getTemperature();
            int maxTokens = chatApiConfig.getMax_tokens();

            // 使用默认配置（如果用户未配置）
            if (apiKey.isEmpty()) {
                apiKey = "sk-e4b872b156584ddb932b767be0bbe938"; // 使用原有项目的 DeepSeek API Key
                System.out.println("使用默认 DeepSeek API Key");
            }
            if (baseUrl.isEmpty()) {
                baseUrl = "https://api.deepseek.com"; // 使用原有项目的 DeepSeek API Base URL
                System.out.println("使用默认 DeepSeek API Base URL: " + baseUrl);
            }
            if (model.isEmpty()) {
                model = "deepseek-chat"; // 默认 DeepSeek 模型
                System.out.println("使用默认 DeepSeek 模型: " + model);
            }

            System.out.println("准备调用 AI API: ");
            System.out.println("- Model: " + model);
            System.out.println("- Base URL: " + baseUrl);
            System.out.println("- API Key: " + apiKey.substring(0, 5) + "..." + apiKey.substring(apiKey.length() - 5));
            System.out.println("- User Input: " + userInput);

            // 构建请求数据
            Map<String, Object> requestData = new HashMap<>();
            requestData.put("model", model);
            requestData.put("temperature", temperature);
            requestData.put("max_tokens", maxTokens);

            Map<String, String> systemMessage = new HashMap<>();
            systemMessage.put("role", "system");
            systemMessage.put("content", buildSystemPrompt());

            Map<String, String> userMessage = new HashMap<>();
            userMessage.put("role", "user");
            userMessage.put("content", userInput);

            requestData.put("messages", new Object[] { systemMessage, userMessage });

            // 发送请求到AI API
            System.out.println("发送请求到 AI API...");
            String response = sendApiRequest(baseUrl, apiKey, requestData);
            System.out.println("收到 API 响应: " + response.substring(0, Math.min(200, response.length())) + "...");

            // 解析响应
            String parsedResponse = parseApiResponse(response);
            System.out.println("解析后的响应: " + parsedResponse);
            return parsedResponse;
        } catch (Exception e) {
            System.out.println("AI 聊天 API 调用失败: " + e.getMessage());
            e.printStackTrace();
            return "抱歉，我暂时无法回答你的问题。错误原因: " + e.getMessage();
        }
    }

    private String buildSystemPrompt() {
        ConfigManager.Config config = configManager.getConfig();
        ConfigManager.PersonalityConfig personality = config.getPersonality();

        StringBuilder prompt = new StringBuilder();
        prompt.append("你是一个VTuber助手，名为").append(personality.getName()).append("。\n");
        prompt.append("你的性格特点：").append(personality.getDescription()).append("\n");
        prompt.append("你的语气风格：").append(personality.getTone()).append("\n");
        prompt.append("你的问候语：").append(personality.getGreeting()).append("\n");
        prompt.append("请以友好、活泼的方式与用户交流，回答用户的问题。");

        return prompt.toString();
    }

    private String sendApiRequest(String baseUrl, String apiKey, Map<String, Object> requestData) throws Exception {
        // 构建完整的 API 端点 URL
        String apiUrl = baseUrl;
        if (!apiUrl.endsWith("/chat/completions")) {
            apiUrl = apiUrl + (apiUrl.endsWith("/") ? "" : "/") + "chat/completions";
        }

        System.out.println("API 端点 URL: " + apiUrl);

        URL url = new URL(apiUrl);
        HttpURLConnection connection = (HttpURLConnection) url.openConnection();
        connection.setRequestMethod("POST");
        connection.setRequestProperty("Content-Type", "application/json");
        connection.setRequestProperty("Authorization", "Bearer " + apiKey);
        connection.setDoOutput(true);
        connection.setConnectTimeout(30000); // 30秒连接超时
        connection.setReadTimeout(30000); // 30秒读取超时

        // 写入请求体
        try (OutputStream os = connection.getOutputStream()) {
            byte[] input = objectMapper.writeValueAsString(requestData).getBytes(StandardCharsets.UTF_8);
            System.out.println("请求体大小: " + input.length + " 字节");
            System.out.println("请求体内容: " + objectMapper.writeValueAsString(requestData));
            os.write(input, 0, input.length);
        }

        // 获取响应状态码
        int responseCode = connection.getResponseCode();
        System.out.println("API 响应状态码: " + responseCode + " " + connection.getResponseMessage());

        // 读取响应
        StringBuilder response = new StringBuilder();
        if (responseCode == HttpURLConnection.HTTP_OK) {
            // 成功响应
            try (BufferedReader br = new BufferedReader(
                    new InputStreamReader(connection.getInputStream(), StandardCharsets.UTF_8))) {
                String responseLine;
                while ((responseLine = br.readLine()) != null) {
                    response.append(responseLine.trim());
                }
            }
        } else {
            // 错误响应
            try (BufferedReader br = new BufferedReader(
                    new InputStreamReader(connection.getErrorStream(), StandardCharsets.UTF_8))) {
                String responseLine;
                while ((responseLine = br.readLine()) != null) {
                    response.append(responseLine.trim());
                }
            }
            System.out.println("API 错误响应: " + response.toString());
            throw new Exception("API 请求失败，状态码: " + responseCode + "，响应: " + response.toString());
        }

        connection.disconnect();
        return response.toString();
    }

    private String parseApiResponse(String response) throws Exception {
        Map<String, Object> responseMap = objectMapper.readValue(response, Map.class);
        if (responseMap.containsKey("choices")) {
            Object choicesObj = responseMap.get("choices");
            if (choicesObj instanceof java.util.ArrayList) {
                java.util.ArrayList<?> choicesList = (java.util.ArrayList<?>) choicesObj;
                if (!choicesList.isEmpty()) {
                    Object choiceObj = choicesList.get(0);
                    if (choiceObj instanceof Map) {
                        Map<?, ?> choice = (Map<?, ?>) choiceObj;
                        Object messageObj = choice.get("message");
                        if (messageObj instanceof Map) {
                            Map<?, ?> message = (Map<?, ?>) messageObj;
                            Object contentObj = message.get("content");
                            if (contentObj instanceof String) {
                                return (String) contentObj;
                            }
                        }
                    }
                }
            }
        }
        return "无法解析API响应。";
    }
}
